core/dbt/task/test.py

```diff
@@ -79,37 +92,52 @@ def execute_test(
         table = result['table']
         num_rows = len(table.rows)
         if num_rows != 1:
-            num_cols = len(table.columns)
-            # since we just wrapped our query in `select count(*)`, we are in
-            # big trouble!
             raise InternalException(
                 f"dbt internally failed to execute {test.unique_id}: "
-                f"Returned {num_rows} rows and {num_cols} cols, but expected "
-                f"1 row and 1 column"
+                f"Returned {num_rows} rows, but expected "
+                f"1 row"
+            )
+        num_cols = len(table.columns)
+        if num_cols != 5:
+            raise InternalException(
+                f"dbt internally failed to execute {test.unique_id}: "
+                f"Returned {num_cols} columns, but expected "
+                f"5 columns"
             )
-        return int(table[0][0])
+
+        test_result_data: PrimitiveDict = dict(zip(table.column_names, table.rows[0]))
+
+        return TestResult.from_dict(test_result_data)
 
     def execute(self, test: CompiledTestNode, manifest: Manifest):
-        failed_rows = self.execute_test(test, manifest)
+        result = self.execute_test(test, manifest)
 
         severity = test.config.severity.upper()
         thread_id = threading.current_thread().name
+        num_errors = utils.pluralize(result.validation_errors, 'result')
         status = None
-        if failed_rows == 0:
-            status = TestStatus.Pass
-        elif severity == 'ERROR' or flags.WARN_ERROR:
+        message = None
+        if severity == "ERROR" and result.should_error:
             status = TestStatus.Fail
+            message = f'Got {num_errors}, configured to fail if {result.error_if}'
+        elif result.should_warn:
+            if flags.WARN_ERROR:
+                status = TestStatus.Fail
+                message = f'Got {num_errors}, configured to fail if {result.warn_if}'
+            else:
+                status = TestStatus.Warn
+                message = f'Got {num_errors}, configured to warn if {result.warn_if}'
         else:
-            status = TestStatus.Warn
+            status = TestStatus.Pass
 
         return RunResult(
             node=test,
             status=status,
             timing=[],
             thread_id=thread_id,
             execution_time=0,
-            message=int(failed_rows),
-            adapter_response={}
+            message=message,
```

Comment:
I am _all for_ making the message an actual message. There are some places in our test suite that expect `message` to be an integer. Could we add a new property, `failures`, that can be that integer value? This would also represent a breaking artifact change (`run_results.json`) that we should communicate

It _could be_ the same value as in `adapter_response.rows_affected`, but I'd rather have this be a first-class test property, and a thing that's consistent across adapters.